hubploy/auth.py

```diff
@@ -69,49 +69,65 @@ def registry_auth_gcloud(deployment, project, service_key):
-    )
+    if not service_key and not ecr_role:
+        raise Exception('AWS authentication requires either a service key or the use of a role')
 
-    if not os.path.isfile(service_key_path):
-        raise FileNotFoundError(
-            f'The service_key file {service_key_path} does not exist')
+    if service_key:
+        # Get credentials from standard location
+        service_key_path = os.path.join(
+            'deployments', deployment, 'secrets', service_key
+        )
 
-    original_credential_file_loc = os.environ.get("AWS_SHARED_CREDENTIALS_FILE", None)
+        if not os.path.isfile(service_key_path):
+            raise FileNotFoundError(
+                f'The service_key file {service_key_path} does not exist')
+
+        original_credential_file_loc = os.environ.get("AWS_SHARED_CREDENTIALS_FILE", None)
 
-    try:
         # Set env variable for credential file location
         os.environ["AWS_SHARED_CREDENTIALS_FILE"] = service_key_path
 
-        registry = f'{project}.dkr.ecr.{zone}.amazonaws.com'
-        # Requires amazon-ecr-credential-helper to already be installed
-        # this adds necessary line to authenticate docker with ecr
-        docker_config_dir = os.path.expanduser('~/.docker')
-        os.makedirs(docker_config_dir, exist_ok=True)
-        docker_config = os.path.join(docker_config_dir, 'config.json')
-        if os.path.exists(docker_config):
-            with open(docker_config, 'r') as f:
-                config = json.load(f)
-        else:
-            config = {}
-
-        config.setdefault('credHelpers', {})[registry] = 'ecr-login'
-        with open(docker_config, 'w') as f:
-            json.dump(config, f)
+        try:
+            registry = f'{project}.dkr.ecr.{zone}.amazonaws.com'
+            # TODO: fix this comment
+            # Requires amazon-ecr-credential-helper to already be installed
+            # this adds necessary line to authenticate docker with ecr
+            docker_config_dir = os.path.expanduser('~/.docker')
+            os.makedirs(docker_config_dir, exist_ok=True)
+            docker_config = os.path.join(docker_config_dir, 'config.json')
+            if os.path.exists(docker_config):
+                with open(docker_config, 'r') as f:
+                    config = json.load(f)
+            else:
+                config = {}
+
+            config.setdefault('credHelpers', {})[registry] = 'ecr-login'
+            with open(docker_config, 'w') as f:
+                json.dump(config, f)
+
+            yield
+
+        finally:
+            if service_key:
+                # Unset env variable for credential file location
+                unset_env_var("AWS_SHARED_CREDENTIALS_FILE", original_credential_file_loc)
+
+    if ecr_role:
+        subprocess.check_call([
+            'aws', 'sts', 'assume-role',
+            f'--role-arn={ecr_role}',
+            '--role-session-name=docker'
+        ])
 
         yield
```

Comment:
In general, I try to have only one yield in a context manager. Otherwise it can get hard to reason about where the `with` statement is actually executing. Do you think it's possible to move the conditionals around to make that the case?